pkg/app/piped/livestatereporter/cloudrun/report.go

```diff
@@ -0,0 +1,135 @@
+
+package cloudrun
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+
+	"github.com/pipe-cd/pipecd/pkg/app/piped/livestatestore/cloudrun"
+	"github.com/pipe-cd/pipecd/pkg/app/server/service/pipedservice"
+	"github.com/pipe-cd/pipecd/pkg/config"
+	"github.com/pipe-cd/pipecd/pkg/model"
+)
+
+type applicationLister interface {
+	ListByCloudProvider(name string) []*model.Application
+}
+
+type apiClient interface {
+	ReportApplicationLiveState(ctx context.Context, req *pipedservice.ReportApplicationLiveStateRequest, opts ...grpc.CallOption) (*pipedservice.ReportApplicationLiveStateResponse, error)
+	ReportApplicationLiveStateEvents(ctx context.Context, req *pipedservice.ReportApplicationLiveStateEventsRequest, opts ...grpc.CallOption) (*pipedservice.ReportApplicationLiveStateEventsResponse, error)
+}
+
+type Reporter interface {
+	Run(ctx context.Context) error
+	ProviderName() string
+}
+
+type reporter struct {
+	provider              config.PipedCloudProvider
+	appLister             applicationLister
+	stateGetter           cloudrun.Getter
+	apiClient             apiClient
+	snapshotFlushInterval time.Duration
+	logger                *zap.Logger
+
+	snapshotVersions map[string]model.ApplicationLiveStateVersion
+}
+
+func NewReporter(cp config.PipedCloudProvider, appLister applicationLister, stateGetter cloudrun.Getter, apiClient apiClient, logger *zap.Logger) Reporter {
+	logger = logger.Named("cloudrun-reporter").With(
+		zap.String("cloud-provider", cp.Name),
+	)
+	return &reporter{
+		provider:              cp,
+		appLister:             appLister,
+		stateGetter:           stateGetter,
+		apiClient:             apiClient,
+		snapshotFlushInterval: 10 * time.Minute,
+		logger:                logger,
+		snapshotVersions:      make(map[string]model.ApplicationLiveStateVersion),
+	}
+}
+
+func (r *reporter) Run(ctx context.Context) error {
+	r.logger.Info("start running app live state reporter")
+
+	r.logger.Info("waiting for livestatestore to be ready")
+	if err := r.stateGetter.WaitForReady(ctx, 10*time.Minute); err != nil {
+		r.logger.Error("livestatestore was unable to be ready in time", zap.Error(err))
+		return err
+	}
+
+	snapshotTicker := time.NewTicker(r.snapshotFlushInterval)
+	defer snapshotTicker.Stop()
+
+	for {
+		select {
+		case <-snapshotTicker.C:
+			r.flushSnapshots(ctx)
+
+		case <-ctx.Done():
+			r.logger.Info("app live state reporter has been stopped")
+			return nil
+		}
+	}
+}
+
+func (r *reporter) ProviderName() string {
+	return r.provider.Name
+}
+
+func (r *reporter) flushSnapshots(ctx context.Context) error {
+	apps := r.appLister.ListByCloudProvider(r.provider.Name)
+	for _, app := range apps {
+		state, ok := r.stateGetter.GetState(app.Id)
+		if !ok {
+			r.logger.Info(fmt.Sprintf("no app state of cloudrun application %s to report", app.Id))
+			continue
+		}
+
+		snapshot := &model.ApplicationLiveStateSnapshot{
+			ApplicationId: app.Id,
+			EnvId:         app.EnvId,
```

Comment:
We can remove it since the environment is no longer used.

Reply:
Exactly, Let me apply it.